File: AST_Fast/test-samples/sample.java
package com.example.web3backend;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.web.bind.annotation.*;

/**
 * Sample Java Spring Boot application for Web3 backend services
 * Demonstrates enterprise Java patterns and Web3 integration
 */

@RestController
@RequestMapping("/api/v1")
public class Web3Controller {
    
    @Autowired
    private Web3Service web3Service;
    
    @Autowired
    private DatabaseService databaseService;
    
    @GetMapping("/wallet/{address}")
    public ResponseEntity<WalletInfo> getWalletInfo(@PathVariable String address) {
        try {
            WalletInfo walletInfo = web3Service.getWalletInfo(address);
            return ResponseEntity.ok(walletInfo);
        } catch (Exception e) {
            return ResponseEntity.badRequest().build();
        }
    }
    
    @PostMapping("/transaction")
    public ResponseEntity<TransactionResult> sendTransaction(@RequestBody TransactionRequest request) {
        try {
            TransactionResult result = web3Service.sendTransaction(request);
            return ResponseEntity.ok(result);
        } catch (Exception e) {
            return ResponseEntity.badRequest().build();
        }
    }
}

@Service
public class Web3Service {
    
    private static final String DEFAULT_RPC_URL = "https://mainnet.infura.io/v3/YOUR-PROJECT-ID";
    private static final BigInteger DEFAULT_GAS_LIMIT = BigInteger.valueOf(21000);
    
    @Autowired
    private CacheService cacheService;
    
    private final Map<String, ContractMetadata> contractCache = new ConcurrentHashMap<>();
    
    public WalletInfo getWalletInfo(String address) throws Web3Exception {
        validateAddress(address);
        
        // Check cache first
        String cacheKey = "wallet_" + address;
        WalletInfo cached = cacheService.get(cacheKey, WalletInfo.class);
        if (cached != null) {
            return cached;
        }
        
        // Fetch from blockchain
        BigInteger balance = getBalance(address);
        List<TokenBalance> tokenBalances = getTokenBalances(address);
        
        WalletInfo walletInfo = new WalletInfo(address, balance, tokenBalances);
        
        // Cache the result
        cacheService.put(cacheKey, walletInfo, 300); // 5 minutes TTL
        
        return walletInfo;
    }
    
    public TransactionResult sendTransaction(TransactionRequest request) throws Web3Exception {
        validateTransactionRequest(request);
        
        // Estimate gas
        BigInteger gasEstimate = estimateGas(request);
        
        // Build transaction
        Transaction transaction = buildTransaction(request, gasEstimate);
        
        // Sign and send
        String txHash = signAndSendTransaction(transaction);
        
        return new TransactionResult(txHash, transaction.getGasLimit(), true);
    }
    
    private BigInteger getBalance(String address) throws Web3Exception {
        try {
            // Simulated Web3 call
            return BigInteger.valueOf(1000000000000000000L); // 1 ETH
        } catch (Exception e) {
            throw new Web3Exception("Failed to get balance", e);
        }
    }
    
    private List<TokenBalance> getTokenBalances(String address) throws Web3Exception {
        List<TokenBalance> balances = new ArrayList<>();
        
        // Get popular tokens
        String[] popularTokens = {
            "0xA0b86a33E6441b8435B3b5a47c3bBc1e7d5b2e0f", // USDC
            "0xdAC17F958D2ee523a2206206994597C13D831ec7", // USDT
            "0x6B175474E89094C44Da98b954EedeAC495271d0F"  // DAI
        };
        
        for (String tokenAddress : popularTokens) {
            try {
                BigInteger balance = getTokenBalance(address, tokenAddress);
                ContractMetadata metadata = getContractMetadata(tokenAddress);
                balances.add(new TokenBalance(tokenAddress, balance, metadata));
            } catch (Exception e) {
                // Log error but continue with other tokens
                System.err.println("Failed to get balance for token: " + tokenAddress);
            }
        }
        
        return balances;
    }
    
    private BigInteger getTokenBalance(String walletAddress, String tokenAddress) throws Web3Exception {
        // Simulated ERC20 balanceOf call
        return BigInteger.valueOf(1000000); // 1M tokens (assuming 6 decimals)
    }
    
    private ContractMetadata getContractMetadata(String contractAddress) throws Web3Exception {
        // Check cache first
        if (contractCache.containsKey(contractAddress)) {
            return contractCache.get(contractAddress);
        }
        
        // Fetch metadata from contract
        ContractMetadata metadata = fetchContractMetadata(contractAddress);
        contractCache.put(contractAddress, metadata);
        
        return metadata;
    }
    
    private ContractMetadata fetchContractMetadata(String contractAddress) throws Web3Exception {
        // Simulated contract calls for name, symbol, decimals
        return new ContractMetadata(
            "Sample Token",
            "SAMPLE",
            18,
            BigInteger.valueOf(1000000000)
        );
    }
    
    private void validateAddress(String address) throws Web3Exception {
        if (address == null || !address.matches("^0x[a-fA-F0-9]{40}$")) {
            throw new Web3Exception("Invalid Ethereum address: " + address);
        }
    }
    
    private void validateTransactionRequest(TransactionRequest request) throws Web3Exception {
        if (request == null) {
            throw new Web3Exception("Transaction request cannot be null");
        }
        
        validateAddress(request.getTo());
        
        if (request.getAmount().compareTo(BigInteger.ZERO) <= 0) {
            throw new Web3Exception("Transaction amount must be positive");
        }
    }
    
    private BigInteger estimateGas(TransactionRequest request) throws Web3Exception {
        // Simulated gas estimation
        return DEFAULT_GAS_LIMIT;
    }
    
    private Transaction buildTransaction(TransactionRequest request, BigInteger gasLimit) {
        return new Transaction(
            request.getTo(),
            request.getAmount(),
            gasLimit,
            request.getGasPrice(),
            request.getData()
        );
    }
    
    private String signAndSendTransaction(Transaction transaction) throws Web3Exception {
        // Simulated transaction signing and sending
        return "0x" + "a".repeat(64); // Mock transaction hash
    }
}

@Service
public class DatabaseService {
    
    private final Map<String, Object> mockDatabase = new ConcurrentHashMap<>();
    
    public void saveWalletActivity(String address, String activity) {
        String key = "activity_" + address;
        List<String> activities = (List<String>) mockDatabase.getOrDefault(key, new ArrayList<String>());
        activities.add(activity);
        mockDatabase.put(key, activities);
    }
    
    public List<String> getWalletActivity(String address) {
        String key = "activity_" + address;
        return (List<String>) mockDatabase.getOrDefault(key, new ArrayList<String>());
    }
    
    public void saveTransaction(TransactionRecord record) {
        mockDatabase.put("tx_" + record.getHash(), record);
    }
    
    public Optional<TransactionRecord> getTransaction(String hash) {
        TransactionRecord record = (TransactionRecord) mockDatabase.get("tx_" + hash);
        return Optional.ofNullable(record);
    }
}

@Service
public class CacheService {
    
    private final Map<String, CacheEntry> cache = new ConcurrentHashMap<>();
    private static final long DEFAULT_TTL_SECONDS = 300; // 5 minutes
    
    public <T> void put(String key, T value, long ttlSeconds) {
        long expiryTime = System.currentTimeMillis() + (ttlSeconds * 1000);
        cache.put(key, new CacheEntry(value, expiryTime));
    }
    
    public <T> T get(String key, Class<T> type) {
        CacheEntry entry = cache.get(key);
        
        if (entry == null) {
            return null;
        }
        
        if (entry.isExpired()) {
            cache.remove(key);
            return null;
        }
        
        return type.cast(entry.getValue());
    }
    
    public void evict(String key) {
        cache.remove(key);
    }
    
    public void clear() {
        cache.clear();
    }
    
    private static class CacheEntry {
        private final Object value;
        private final long expiryTime;
        
        public CacheEntry(Object value, long expiryTime) {
            this.value = value;
            this.expiryTime = expiryTime;
        }
        
        public Object getValue() {
            return value;
        }
        
        public boolean isExpired() {
            return System.currentTimeMillis() > expiryTime;
        }
    }
}

// Data classes
public class WalletInfo {
    private final String address;
    private final BigInteger ethBalance;
    private final List<TokenBalance> tokenBalances;
    private final long timestamp;
    
    public WalletInfo(String address, BigInteger ethBalance, List<TokenBalance> tokenBalances) {
        this.address = address;
        this.ethBalance = ethBalance;
        this.tokenBalances = tokenBalances;
        this.timestamp = System.currentTimeMillis();
    }
    
    // Getters
    public String getAddress() { return address; }
    public BigInteger getEthBalance() { return ethBalance; }
    public List<TokenBalance> getTokenBalances() { return tokenBalances; }
    public long getTimestamp() { return timestamp; }
}

public class TokenBalance {
    private final String contractAddress;
    private final BigInteger balance;
    private final ContractMetadata metadata;
    
    public TokenBalance(String contractAddress, BigInteger balance, ContractMetadata metadata) {
        this.contractAddress = contractAddress;
        this.balance = balance;
        this.metadata = metadata;
    }
    
    // Getters
    public String getContractAddress() { return contractAddress; }
    public BigInteger getBalance() { return balance; }
    public ContractMetadata getMetadata() { return metadata; }
}

public class ContractMetadata {
    private final String name;
    private final String symbol;
    private final int decimals;
    private final BigInteger totalSupply;
    
    public ContractMetadata(String name, String symbol, int decimals, BigInteger totalSupply) {
        this.name = name;
        this.symbol = symbol;
        this.decimals = decimals;
        this.totalSupply = totalSupply;
    }
    
    // Getters
    public String getName() { return name; }
    public String getSymbol() { return symbol; }
    public int getDecimals() { return decimals; }
    public BigInteger getTotalSupply() { return totalSupply; }
}

public class TransactionRequest {
    private String to;
    private BigInteger amount;
    private BigInteger gasPrice;
    private byte[] data;
    
    // Constructors, getters, and setters
    public TransactionRequest() {}
    
    public String getTo() { return to; }
    public void setTo(String to) { this.to = to; }
    
    public BigInteger getAmount() { return amount; }
    public void setAmount(BigInteger amount) { this.amount = amount; }
    
    public BigInteger getGasPrice() { return gasPrice; }
    public void setGasPrice(BigInteger gasPrice) { this.gasPrice = gasPrice; }
    
    public byte[] getData() { return data; }
    public void setData(byte[] data) { this.data = data; }
}

public class TransactionResult {
    private final String hash;
    private final BigInteger gasUsed;
    private final boolean success;
    private final long timestamp;
    
    public TransactionResult(String hash, BigInteger gasUsed, boolean success) {
        this.hash = hash;
        this.gasUsed = gasUsed;
        this.success = success;
        this.timestamp = System.currentTimeMillis();
    }
    
    // Getters
    public String getHash() { return hash; }
    public BigInteger getGasUsed() { return gasUsed; }
    public boolean isSuccess() { return success; }
    public long getTimestamp() { return timestamp; }
}

public class Transaction {
    private final String to;
    private final BigInteger amount;
    private final BigInteger gasLimit;
    private final BigInteger gasPrice;
    private final byte[] data;
    
    public Transaction(String to, BigInteger amount, BigInteger gasLimit, BigInteger gasPrice, byte[] data) {
        this.to = to;
        this.amount = amount;
        this.gasLimit = gasLimit;
        this.gasPrice = gasPrice;
        this.data = data;
    }
    
    // Getters
    public String getTo() { return to; }
    public BigInteger getAmount() { return amount; }
    public BigInteger getGasLimit() { return gasLimit; }
    public BigInteger getGasPrice() { return gasPrice; }
    public byte[] getData() { return data; }
}

public class TransactionRecord {
    private final String hash;
    private final String from;
    private final String to;
    private final BigInteger amount;
    private final long timestamp;
    
    public TransactionRecord(String hash, String from, String to, BigInteger amount) {
        this.hash = hash;
        this.from = from;
        this.to = to;
        this.amount = amount;
        this.timestamp = System.currentTimeMillis();
    }
    
    // Getters
    public String getHash() { return hash; }
    public String getFrom() { return from; }
    public String getTo() { return to; }
    public BigInteger getAmount() { return amount; }
    public long getTimestamp() { return timestamp; }
}

// Exception classes
public class Web3Exception extends Exception {
    public Web3Exception(String message) {
        super(message);
    }
    
    public Web3Exception(String message, Throwable cause) {
        super(message, cause);
    }
}

File: AST_Fast/test-samples/sample.go
package main

import (
	"context"
	"fmt"
	"log"
	"math/big"
	
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/core/types"
	"github.com/ethereum/go-ethereum/ethclient"
)

// TokenContract represents an ERC20 token contract
type TokenContract struct {
	Address     common.Address
	Name        string
	Symbol      string
	Decimals    uint8
	TotalSupply *big.Int
}

// WalletService handles wallet operations
type WalletService struct {
	client   *ethclient.Client
	chainID  *big.Int
	gasLimit uint64
}

// TransactionResult contains transaction information
type TransactionResult struct {
	Hash        common.Hash
	BlockNumber uint64
	GasUsed     uint64
	Success     bool
}

// NewWalletService creates a new wallet service instance
func NewWalletService(rpcURL string) (*WalletService, error) {
	client, err := ethclient.Dial(rpcURL)
	if err != nil {
		return nil, fmt.Errorf("failed to connect to Ethereum client: %w", err)
	}

	chainID, err := client.NetworkID(context.Background())
	if err != nil {
		return nil, fmt.Errorf("failed to get network ID: %w", err)
	}

	return &WalletService{
		client:   client,
		chainID:  chainID,
		gasLimit: 21000,
	}, nil
}

// GetBalance returns the ETH balance of an address
func (ws *WalletService) GetBalance(address common.Address) (*big.Int, error) {
	return ws.client.BalanceAt(context.Background(), address, nil)
}

// GetTokenBalance returns the token balance for a specific ERC20 contract
func (ws *WalletService) GetTokenBalance(tokenAddress, walletAddress common.Address) (*big.Int, error) {
	// This is a simplified implementation
	// In reality, you would call the balanceOf method on the token contract
	return big.NewInt(0), nil
}

// SendTransaction sends a transaction to the blockchain
func (ws *WalletService) SendTransaction(tx *types.Transaction) (*TransactionResult, error) {
	err := ws.client.SendTransaction(context.Background(), tx)
	if err != nil {
		return nil, fmt.Errorf("failed to send transaction: %w", err)
	}

	return &TransactionResult{
		Hash:    tx.Hash(),
		Success: true,
	}, nil
}

// EstimateGas estimates the gas required for a transaction
func (ws *WalletService) EstimateGas(from, to common.Address, data []byte) (uint64, error) {
	msg := ethereum.CallMsg{
		From: from,
		To:   &to,
		Data: data,
	}
	
	return ws.client.EstimateGas(context.Background(), msg)
}

// BlockchainMonitor monitors blockchain events
type BlockchainMonitor struct {
	client        *ethclient.Client
	subscriptions map[string]chan interface{}
}

// NewBlockchainMonitor creates a new blockchain monitor
func NewBlockchainMonitor(client *ethclient.Client) *BlockchainMonitor {
	return &BlockchainMonitor{
		client:        client,
		subscriptions: make(map[string]chan interface{}),
	}
}

// SubscribeToBlocks subscribes to new block headers
func (bm *BlockchainMonitor) SubscribeToBlocks() error {
	headers := make(chan *types.Header)
	sub, err := bm.client.SubscribeNewHead(context.Background(), headers)
	if err != nil {
		return err
	}

	go func() {
		for {
			select {
			case err := <-sub.Err():
				log.Printf("Subscription error: %v", err)
				return
			case header := <-headers:
				log.Printf("New block: %d", header.Number)
			}
		}
	}()

	return nil
}

// TokenMetadata holds metadata for a token
var (
	DefaultGasLimit = uint64(21000)
	MaxGasPrice     = big.NewInt(100000000000) // 100 gwei
)

const (
	EthereumMainnet = 1
	EthereumGoerli  = 5
	EthereumSepolia = 11155111
)

func main() {
	service, err := NewWalletService("https://mainnet.infura.io/v3/YOUR-PROJECT-ID")
	if err != nil {
		log.Fatal(err)
	}

	address := common.HexToAddress("0x742d35Cc6634C0532925a3b8D162444568aF8A3c")
	balance, err := service.GetBalance(address)
	if err != nil {
		log.Fatal(err)
	}

	fmt.Printf("Balance: %s ETH\n", balance.String())
}


